design/rest-api-guideline.md

```diff
@@ -0,0 +1,406 @@
+|404|Indicates that the requested resource does not exist|
+|405|Indicates that the method chosen by the client to access a resource is not supported|
+|409|Indicates that the resource the client is trying to create already exists or some conflict when processing the request|
+|415|Indicates that the requested media type is not supported|
+
+#### Server Side Error Codes
+
+|Code|Description|
+|---|---|
+|500|Indicates that the server could not fulfill the request due to some unexpected error|
+
+### Http Verbs
+
+Based on the REST architectural style, HTTP Verbs should be used as the means to communicate to APIs the operations to be performed
+on their resources. While HTTP verbs cover by themselves the most common operations you might want to look the [Controller Resource](#controller-resources) as an alternative on how to overcome any limitation imposed by the usage of HTTP verbs herein defined.
+
+The table below defines the HTTP verbs you should consider when exposing operations through your API and the circumstances where they apply.
+
+|Verb|Description|Idempotent|
+|---|---|---|
+|GET|Requests the current state of a resource|Yes|
+|POST|Creates a new resource or submit a command|No|
+|PUT|Replaces a resource state|Yes|
+|PATCH|Applies partial updates to a resource|Yes|
+|DELETE|Removes a resource|Yes|
+
+#### GET
+
+HTTP `GET` should be used to obtain current state of a resource. It is usually related with operations on resources that return either a list or a single representation, or start some action that does not change the resource state.
+
+##### Expected Status Codes
+
+It is expected that APIs responding to HTTP `GET` requests respect the following HTTP status codes to communicate the result of the operation:
+
+|Status Codes|
+|---|
+|200|
+|204|
+|307|
+|400|
+|401|
+|403|
+|404|
+|415|
+|500|
+
+#### POST
+
+HTTP `POST` should be used to create new resources. It is a non-safe and non-idempotent operation and clients should be aware of that.
+
+Successful `POST` requests should always result in a `201` status code and empty response. The response should also contain a reference
+to the resource that was created:
+
+```
+HTTP/1.1 201 Created
+Content-Length: 0
+Location: http://mykeycloak.com/realms/{realms}/apis/users/{version}/1234
+```
+
+##### Expected Status Codes
+
+It is expected that APIs responding to HTTP `POST` requests respect the following HTTP status codes to communicate the result of the operation:
+
+|Status Codes|
+|---|
+|201|
+|400|
+|401|
+|403|
+|404|
+|409|
+|415|
+|500|
+
+#### PUT
+
+HTTP `PUT` should be used to replace a resource state.
+
+Successful `PUT` requests should always result in a `204` status code and empty response.
+
+```
+HTTP/1.1 204 No Content
+Content-Length: 0
+```
+
+#### PATCH
+
+HTTP `PATCH` should be used to partially update the resource state.
+
+Successful `PATCH` requests should always result in a `204` status code and empty response.
```

Comment:
Again, should the response be empty? It IMHO makes sense to return the final result when doing a partial update.

Reply:
I'm not sure, because it is a very specific operation. I'll leave open so others can give their opinion.